asttokens/mark_tokens.py

```diff
@@ -219,13 +219,39 @@ def visit_subscript(self, node, first_token, last_token):
     return (first_token, self._code.find_token(last_token, token.OP, ']'))
 
   def visit_tuple(self, node, first_token, last_token):
-    # A tuple doesn't include parens; if there is a trailing comma, make it part of the tuple.
-    try:
-      maybe_comma = self._code.next_token(last_token)
-      if util.match_token(maybe_comma, token.OP, ','):
-        last_token = maybe_comma
-    except IndexError:
-      pass
+    # Tuples may or may not contain parentheses
+    # If they do, we want to include them
+    # We may or may not already have them, we need to find out
+    # We have them if the tokens start and end with matching ()
+    # meaning if we remove them we are still left with matching parens
+    contains_parens = False
+    if util.is_parentheses(first_token, last_token):
```

Comment:
I don't think it's OK to decide on the spot this way. A tuple like `((1), (2))` will present itself as `(1), (2)` pre-3.8, and as `((1), (2))` post-3.8, but in both cases we start and end with parentheses. I think  we need to have a version check, and expand to include the parens if pre-3.8. (Or alternatively, skip the paren in >=3.8, if it might be better to keep parens out, to maintain past behavior.)

Reply:
That's handled in the following lines as explained in the preceding comment

Reply:
Sorry, I think I understand your point a bit better now, I need to think about it. It seems overly complicated.